### PR TITLE
Replace all non-test uses of new T[0] with Array.Empty<T>

### PR DESCRIPTION
**Pull request by James-Ko**
I ran a little shell command to find and replace most of the occurrences of `new T[0]` in the repo:

    cd corefx
    grep -rl "new [^ (]*\[0\]" . | grep -vi test | xargs start notepad++

There were about 18 matches, with an additional match I spotted in `Queue<T>.ToArray`.

Additional notes:

 - Removed caching of `EmptySequence` in `XElement` and `XAttribute`; this is already done by `Array.Empty`.
 - The build should fail at first because I forgot to add `using` statements to the classes I changed. Will fix once the compiler tells where it can't recognize `Array`.

### CODE REVIEW

File: src/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentBag.cs [hunk 1]
```diff
@@ -513,7 +513,7 @@ public T[] ToArray()
         {
             // Short path if the bag is empty
             if (_headList == null)
-                return new T[0];
+				return Array.Empty<T>();
```
Comment (stephentoub) [inline src/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentBag.cs]: I explicitly avoided changing this one in the past (and others in ToArray methods) because the docs for it are explicit about it creating a "new" array.

File: src/System.Dynamic.Runtime/src/System/Dynamic/DynamicObject.cs [hunk 2]
```diff
@@ -382,7 +382,7 @@ public override DynamicMetaObject BindDeleteIndex(DeleteIndexBinder binder, Dyna
 
             private delegate DynamicMetaObject Fallback(DynamicMetaObject errorSuggestion);
 
-            private readonly static Expression[] s_noArgs = new Expression[0]; // used in reference comparison, requires unique object identity
+            private readonly static Expression[] s_noArgs = Array.Empty<Expression>(); // used in reference comparison, requires unique object identity
```
Comment (stephentoub) [inline src/System.Dynamic.Runtime/src/System/Dynamic/DynamicObject.cs]: See the comment; this shouldn't be changed.

File: src/System.Reflection.Metadata/src/System/Reflection/Internal/Utilities/EmptyArray.cs [hunk 3]
```diff
@@ -5,6 +5,6 @@ namespace System.Reflection.Internal
 {
     internal static class EmptyArray<T>
     {
-        internal static readonly T[] Instance = new T[0];
+        internal static readonly T[] Instance = Array.Empty<T>();
```
Comment (stephentoub) [inline src/System.Reflection.Metadata/src/System/Reflection/Internal/Utilities/EmptyArray.cs]: Does this work?  I thought this assembly built against an earlier contract that lacks Empty.

File: src/System.Threading.Tasks.Dataflow/src/Internal/ImmutableList.cs [hunk 4]
```diff
@@ -31,7 +31,7 @@ internal sealed class ImmutableList<T> : IEnumerable<T>
         public static ImmutableList<T> Empty { get { return _empty; } }
 
         /// <summary>Initializes the immutable list to be empty.</summary>
-        private ImmutableList() : this(new T[0]) { }
+        private ImmutableList() : this(Array.Empty<T>()) { }
```
Comment (stephentoub) [inline src/System.Threading.Tasks.Dataflow/src/Internal/ImmutableList.cs]: Ditto; does this compile?

**Comment by stephentoub** [comment]
@khdang, pelase take a look at the XmlSerializer changes to confirm whether they're acceptable or not.

@bartonjs, please take a look at the crypto changes to confirm whether they're acceptable or whether the behavioral impact (object identity of the returned object) is problematic.

@VSadov, please take a look at the dynamic, LINQ, etc. stuff.

**Comment by stephentoub** [comment]
> Will fix once the compiler tells where it can't recognize Array

I'm curious, does this mean you're not building locally before submitting?

File: src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray`1.cs [hunk 5]
```diff
@@ -37,7 +37,7 @@ public partial struct ImmutableArray<T> : IReadOnlyList<T>, IList<T>, IEquatable
         /// <summary>
         /// An empty (initialized) instance of <see cref="ImmutableArray{T}"/>.
         /// </summary>
-        public static readonly ImmutableArray<T> Empty = new ImmutableArray<T>(new T[0]);
+        public static readonly ImmutableArray<T> Empty = new ImmutableArray<T>(Array.Empty<T>());
```
Comment (stephentoub) [inline src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray`1.cs]: Does this compile?  I thought we built this library against a contract that lacks Empty.  And since it's being cached into a singleton in a static, caching it separately isn't really warranted.

**Comment by bartonjs** [comment]
I can't imagine anyone needing reference uniqueness from the affected crypto code, so I'm good with the crypto part of the change.

**Comment by stephentoub** [comment]
Great; thanks, Jeremy.

File: src/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentBag.cs (hunk 1 above)
Comment (hackcraft) [inline src/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentBag.cs]: @stephentoub if that's considered important, should we have a test for `Assert.NotSame(new ConcurrentBag<string>().ToArray(), ConcurrentBag<string>().ToArray())`?

Comment (stephentoub) [inline src/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentBag.cs]: @hackcraft, yeah, I think either we should take this change, or add such a test. Same for other collections.

Comment (hackcraft) [inline src/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentBag.cs]: On the flip side. If this change is taken then a test of `Assert.Same(new ConcurrentBag<string>().ToArray(), ConcurrentBag<string>().ToArray())` would mean any change back would at least have to justify why they removed they performance gain by justifying removing the now-breaking test.

Comment (stephentoub) [inline src/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentBag.cs]: Yes, there are already some of both kinds in the repo, for various things (including some that verify this for ToArray, e.g. there's a Queue.ToArray test that verifies not-sameness and which fails with this PR.)

**Comment by khdang** [comment]
The changes in XmlSerializer look good to me.

File: src/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentBag.cs (hunk 1 above)
Comment (James-Ko) [inline src/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentBag.cs]: @stephentoub To stay consistent, then, I'll remove the changes to Queue.

File: src/System.Dynamic.Runtime/src/System/Dynamic/DynamicObject.cs (hunk 2 above)
Comment (James-Ko) [inline src/System.Dynamic.Runtime/src/System/Dynamic/DynamicObject.cs]: Thanks for pointing that out; this has been fixed in the next commit.

File: src/System.Linq/src/System/Linq/Enumerable.cs [hunk 6]
```diff
@@ -3159,7 +3159,7 @@ internal Buffer(IEnumerable<TElement> source, bool queryInterfaces = true)
 
         internal TElement[] ToArray()
         {
-            if (count == 0) return new TElement[0];
+            if (count == 0) return Array.Empty<TElement>();
```
Comment (VSadov) [inline src/System.Linq/src/System/Linq/Enumerable.cs]: I think we considered changing this, but were not sure if it is enough benefit compared to potential for breaking code relying on distinct array instances. 
User might already have some allocations when hitting this path so one less empty array might not make a lot of improvement in typical code. The documentation says "An IEnumerable<T> to create an array from." If taken literally it seems to say that array is always created.
I wonder if there is a way to estimate if there is actual usage of this with expectation that array is created. Perhaps it can be done by scanning some publicly available code or something...

I think we should not change this just yet, but may worth adding a comment that it is being considered.

**Comment by hackcraft** [comment]
I can see a reverse danger to this; someone could come to expect that the result of some of these operations are all the same object and then be surprised down the line (or perhaps upon deserialisation if they hadn't previously considered this).

I don't see that being a big issue (I can remember people being confused by the fact that `ReferenceEquals(string.Empty, "")` changed between framework versions, but not stymied by it). But better to consider it and rule it out as a problem than not consider it.

**Comment by VSadov** [comment]
LINQ/Dynamic look ok except for ToArray() ones. 

They share the same concern that benefits might not be that great while user might expect new instances and therefore we were reluctant to change these. I wonder what would be the right forum to discuss the ToArray issue in general and to conclude one way or another, possibly in the context of the whole framework so it would be consistent.

File: src/System.Linq/src/System/Linq/Enumerable.cs (hunk 6 above)
Comment (James-Ko) [inline src/System.Linq/src/System/Linq/Enumerable.cs]: Great idea; I agree, many of the places where this is used (`Iterator.ToArray`) seem to imply that they are returning a different array. I'll revert and add a comment on the side.

**Comment by James-Ko** [comment]
@hackcraft Hm. I'm assuming you mean scenarios like this:

    if (queue.ToArray() == Array.Empty<T>())
    {
        // ...

That might happen if someone did stumbled on this by chance and assumed that it would always be true, but in general I think most people would go for checking the length of the array or some variant.

That said, I am undoing the changes I made to Queue for now.

**Comment by krwq** [comment]
Not sure if I remember correctly: I believe we used to have Array.Empty<T> internally some time ago for a while and then we reverted the changes as Array.Empty<T> is not available on some earlier versions of full .NET and we wanted to be compatible with as many versions as possible.

We should find out who did the internal revert to figure out if we should re-do it.

**Comment by stephentoub** [comment]
@krwq, can you do that? Thanks.

File: src/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentBag.cs [hunk 7]
```diff
@@ -513,7 +513,7 @@ public T[] ToArray()
         {
             // Short path if the bag is empty
             if (_headList == null)
-                return new T[0];
+				return new T[0];
```
Comment (stephentoub) [inline src/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentBag.cs]: Please undo this one in the ToArray method.

File: src/System.Collections/src/System/Collections/Generic/Queue.cs [hunk 8]
```diff
@@ -309,8 +309,8 @@ public T[] ToArray()
         {
             T[] arr = new T[_size];
             if (_size == 0)
-                return arr;
-
+                return arr; // consider replacing with Array.Empty<T>() to be consistent with non-generic Queue
+
```
Comment (stephentoub) [inline src/System.Collections/src/System/Collections/Generic/Queue.cs]: Hmm, it looks like I'm being inconsistent with myself.  I'd previously changed ```Queue.ToArray``` to return ```Array.Empty<object>```, doing so for the other NonGeneric collections as well, but I explicitly avoided doing so for the generic ones.  I don't remember now what my rationale was for thinking it was ok to convert one set but not the other.  Seems like (separate from this PR), we should reevaluate that; I'm tempted to roll back those changes to the non-generic collections, as they exist primarily for backwards compatibility, and any code more sensitive to performance should be using the generic ons, especially with value types.

File: src/System.Xml.XDocument/src/System/Linq/Enumerable.cs [hunk 9]
```diff
@@ -192,7 +192,7 @@ internal Buffer(IEnumerable<TElement> source)
 
         internal TElement[] ToArray()
         {
-            if (count == 0) return new TElement[0];
+            if (count == 0) return Array.Empty<TElement>();
```
Comment (stephentoub) [inline src/System.Xml.XDocument/src/System/Linq/Enumerable.cs]: Please revert this.

File: src/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentBag.cs (hunk 7 above)
Comment (stephentoub) [inline src/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentBag.cs]: Nit: unnecessary whitespace changes

**Comment by krwq** [comment]
XML parth looks good to me :+1: Also please disregard my previous comment.

**Comment by stephentoub** [comment]
Thanks, @James-Ko.  Looks good.  Can you please squash this?

**Comment by James-Ko** [comment]
I have no idea why, but it seems that the build is suddenly failing after I squashed the commits. According to [Jenkins](http://dotnet-ci.cloudapp.net/job/dotnet_corefx_linux_debug_tst_prtest/1186/testReport/), it seems to be a problem with `EtwTests` on [Linux](http://dotnet-ci.cloudapp.net/job/dotnet_corefx_linux_debug_tst_prtest/), which I have not heard of nor touched. Should I try a soft reset from the last commit and re-push?

**Comment by stephentoub** [comment]
@James-Ko, I don't believe this is related to your PR.  Others are failing with the same issue.  I'm suspecting that there's either something wrong with the infrastructure or that something's changed with the CoreCLR we're pulling in.  @mmitche, @ellismg, ideas?

**Comment by mmitche** [comment]
This looks like maybe a change got merged in that accidentally enabled etw tests on Linux?  It doesn't makes sense to run them there anyway.

**Comment by stephentoub** [comment]
> it doesn't makes sense to run them there anyway.

It does, actually. These have been running on Linux for a long time. EventSource works on Linux, it just isn't hooked up to ETW. EventListeners still receive events, and these tests make sure those events are raised.

**Comment by ellismg** [comment]
Still trying to dig into this but I have to run for a bit so it won't happen until later tonight.

**Comment by stephentoub** [comment]
@ellismg, I'm looking as well.  I have a hunch.  @mmitche, you recently brought new build tools into CoreCLR, right?  Is it possible that in doing so we switched something related to the $(OS) / $(OSGroup) properties and what defines get set as a result?  It looks like some Windows-specific code that's protected by FEATURE_MANAGED_ETW is getting output in the Linux mscorlib now.

**Comment by stephentoub** [comment]
I think this fixes it: https://github.com/dotnet/coreclr/pull/1246

**Comment by stephentoub** [comment]
LGTM

**Comment by Maxwe11** [comment]
@dotnet-bot test this please
